### PR TITLE
feat(moderation): dehoist command

### LINKED ISSUE #654: todo: Dehoisting.
`s!dehoist`, renaming anybody whose nickname/username hoist (lower code than 'A') to something, get creative on that -Kyra

### PR DESCRIPTION
**Pull request by cfanoulis**
implementation for issue #654

closes #654 

Road to contributor: 5/6 PRs :poggers:

### CODE REVIEW

File: src/commands/Moderation/dehoist.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		});
+	}
+
+	public async run(message: KlasaMessage) {
+		let counter = 0;
+		const promises = [];
+		const members = message.guild!.memberTags;
+		const msg = await message.sendLocale('COMMAND_HIGHERLOWER_LOADING');
+
+		for (const [memberId, memberTag] of members) {
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: Rebase and change this to:

```suggestion
		for (const [memberId, memberTag] of members.manageableMembers()) {
```

As added in https://github.com/kyranet/Skyra/commit/ac47e83863817ddad7d235821ff8d5d723a7862e

File: src/commands/Moderation/dehoist.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			if (memberTag.nickname && memberTag.nickname.charCodeAt(0) < 65 /* That's A */) {
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: That `65` should be a constant like:
```typescript
private readonly kLowestCode = 'A'.charCodeAt(0);
```

So we do this:
```suggestion
			if (memberTag.nickname && memberTag.nickname.charCodeAt(0) < this.kLowestCode) {
```

File: src/commands/Moderation/dehoist.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+				// Replace the first character of the offending user's with a downwards arrow, bringing'em down, down ,down
+				const newNickname = memberTag.nickname!.replace(String.fromCharCode(memberTag.nickname.charCodeAt(0)), '🠷');
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: Please don't repeat the coinflip story again, lmao.
```suggestion
				const newNickname = `🠷${memberTag.nickname.slice(1)}`;
```

File: src/commands/Moderation/dehoist.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+			permissionLevel: PermissionLevels.Moderator,
+			requiredPermissions: ['MANAGE_NICKNAMES']
+		});
+	}
+
+	public async run(message: KlasaMessage) {
+		let counter = 0;
+		const promises = [];
+		const members = message.guild!.memberTags;
+		const msg = await message.sendLocale('COMMAND_HIGHERLOWER_LOADING');
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]:
```suggestion
		const msg = await message.sendLocale('SYSTEM_LOADING');
```

Comment (kyranet) [reply src/commands/Moderation/dehoist.ts]: Also rename `msg` to `response`.

File: src/commands/Moderation/dehoist.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+					.members(memberId)
+					.patch({ data: { nick: newNickname }, reason: 'Dehoisting' }));
+
+				counter++;
+			}
+		}
+		await Promise.all(promises);
+
+		const { TITLE: generateTitle, DESCRIPTION_NOONE, DESCRIPTION: generateDescription } = message.language.tget('COMMAND_DEHOIST_EMBED');
+		await msg.edit(new MessageEmbed()
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: This should return.

File: src/commands/Moderation/dehoist.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+					.guilds(message.guild!.id)
+					.members(memberId)
+					.patch({ data: { nick: newNickname }, reason: 'Dehoisting' }));
+
+				counter++;
+			}
+		}
+		await Promise.all(promises);
+
+		const { TITLE: generateTitle, DESCRIPTION_NOONE, DESCRIPTION: generateDescription } = message.language.tget('COMMAND_DEHOIST_EMBED');
```
Comment (kyranet) [inline src/commands/Moderation/dehoist.ts]: Please don't do this, you kill an innocent kitten by doing this, save the kittens.
image

File: src/languages/en-US.ts [hunk 7]
```diff
@@ -1381,6 +1381,15 @@ export default class extends Language {
 					a temporary ban.`,
 			examples: ['@Pete', '@Pete Spamming all channels.', '@Pete Spamming all channels, for 24 hours.']
 		}),
+		COMMAND_DEHOIST_DESCRIPTION: 'Shoot everyone with the dehoist-inator 3000',
+		COMMAND_DEHOIST_EXTENDED: builder.display('dehoist', {
+			extendedHelp: `The act of hoisting involves adding a character with a value of 64 and below in front of your nickname
```
Comment (kyranet) [inline src/languages/en-US.ts]: How bold of you to assume everyone knows what `64` is.
And much less what it represents in ASCII and why is it relevant.

File: src/languages/en-US.ts [hunk 8]
```diff
@@ -1381,6 +1381,15 @@ export default class extends Language {
...
+			in order to appear higher in the members list. This command replace's with a special character that drags them to the bottom
+			of the list. Ha, get a load of this, mr. \`! !\`
+
```
Comment (kyranet) [inline src/languages/en-US.ts]: And this new line?

File: src/languages/en-US.ts [hunk 9]
```diff
@@ -1381,6 +1381,15 @@ export default class extends Language {
 					a temporary ban.`,
 			examples: ['@Pete', '@Pete Spamming all channels.', '@Pete Spamming all channels, for 24 hours.']
 		}),
+		COMMAND_DEHOIST_DESCRIPTION: 'Shoot everyone with the dehoist-inator 3000',
+		COMMAND_DEHOIST_EXTENDED: builder.display('dehoist', {
+			extendedHelp: `The act of hoisting involves adding a character with a value of 64 and below in front of your nickname
+			in order to appear higher in the members list. This command replace's with a special character that drags them to the bottom
+			of the list. Ha, get a load of this, mr. \`! !\`
+			
+			This command requires **${PERMS.MANAGE_NICKNAMES}**, and only members with lower role hierarchy position
```
Comment (kyranet) [inline src/languages/en-US.ts]: This should go to the `reminder` option from the builder, not in `extendedHelp`.

File: src/languages/en-US.ts [hunk 10]
```diff
@@ -2839,6 +2848,12 @@ export default class extends Language {
 			: `The cooldown for this channel has been set to ${duration(cooldown)}.`,
 		COMMAND_SLOWMODE_TOO_LONG: `${REDCROSS} The maximum amount of time you can set is 6 hours.`,
 		COMMAND_BAN_NOT_BANNABLE: 'The target is not bannable for me.',
+		COMMAND_DEHOIST_LOADING: `${LOADING} Dehoisting has started. This may take a bit...`,
+		COMMAND_DEHOIST_EMBED: {
+			TITLE: users => `Finished dehoisting ${users}`,
+			DESCRIPTION_NOONE: 'No members were dehoisted. A round of applause for your law-abiding users!',
+			DESCRIPTION: users => `${users} members were dehoisted`
```
Comment (kyranet) [inline src/languages/en-US.ts]: 1 member**s**.

- **kyra**: Triggered.
- **vlad**: Triggered.
- **Favna**: Triggered.

File: src/languages/es-ES.ts [hunk 11]
```diff
@@ -1367,6 +1367,15 @@ export default class extends Language {
 					a temporary ban.`,
 			examples: ['@Pete', '@Pete Spamming all channels.', '@Pete Spamming all channels, for 24 hours.']
 		}),
+		COMMAND_DEHOIST_DESCRIPTION: 'Convierta a todos a no izados',
```
Comment (kyranet) [inline src/languages/es-ES.ts]: image

File: src/languages/es-ES.ts [hunk 12]
```diff
@@ -1367,6 +1367,15 @@ export default class extends Language {
...
+		COMMAND_DEHOIST_EXTENDED: builder.display('dehoist', {
+			extendedHelp: `El acto de izar implica agregar un personaje con un valor de 64 y menos delante de su apodo para
+				aparecer más arriba en la lista de miembros.
+				Este comando reemplaza con un carácter especial que los arrastra al final de la lista.
+				Ja, obtener un montón de esto, señor \`! !\`
```
Comment (kyranet) [inline src/languages/es-ES.ts]: Sincerely, I can't think of a way to translate this line to Spanish and make it not sound offensive.

File: src/languages/es-ES.ts [hunk 13]
```diff
@@ -2819,6 +2828,12 @@ export default class extends Language {
 			: `The cooldown for this channel has been set to ${duration(cooldown)}.`,
 		COMMAND_SLOWMODE_TOO_LONG: `${REDCROSS} The maximum amount of time you can set is 6 hours.`,
 		COMMAND_BAN_NOT_BANNABLE: 'The target is not bannable for me.',
+		COMMAND_DEHOIST_LOADING: `${LOADING} La conversion ha comenzado. Esto puede tardar un rato...`,
+		COMMAND_DEHOIST_EMBED: {
+			TITLE: users => `Termine de convertir ${users} miembros a no izados`,
+			DESCRIPTION_NOONE: 'Ningun miembro se convirtio en no izado. Una ronda de aplausos para sus miembros que respetan las reglas!',
+			DESCRIPTION: users => `${users} miembros se volvieron no izados`
```
Comment (kyranet) [inline src/languages/es-ES.ts]: 1 miembro**s**

- **everyone**: Triggered.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/languages/es-ES.ts (hunk 11 above)
Comment (kyranet) [reply src/languages/es-ES.ts]: If you **really** want to do Spanish translations, at very least, use this. https://www.wordreference.com/es/translation.asp?tranword=hoist

File: src/commands/Moderation/dehoist.ts (hunk 6 above)
Comment (kyranet) [reply src/commands/Moderation/dehoist.ts]: fr, save the kittens.
image

Comment (cfanoulis) [reply src/commands/Moderation/dehoist.ts]: KILL THE KITTENS! (im joking, @peta dont kill me)

**Comment by cfanoulis** [comment]
skipping Spanish translations because I doooon't know that much Spanish

**Review by kyranet — APPROVED** [review]

**Pull request merged**